### ISSUE TITLE
Using a trained model on GPU or CPU backend results in different outputs

### ISSUE THREAD

**Issue by sorensf** [deeplearning4j/deeplearning4j#4688]
#### Issue Description
Having trained a CNN model and saving it to disk. Then when the model is loaded a CPU backend it outputs one probability distribution and when loaded on a GPU backend it outputs a different probability distribution.

l using the CNN configuration https://gist.github.com/sorensf/88b95d32018a204daca197f151b142c3 
CPU backend is builts as a .jar using the pom.xml https://gist.github.com/sorensf/a7dc3fd62345246ad420a5aed3eef2f9
GPU backend is builts as a .jar using the pom.xml https://gist.github.com/sorensf/ed27aa37d0de421c0953eeefbf6ef92a

#### Version Information
Dl4j/ND4j: Version 0.9.1
Platform: Tested on Windows 7 and Linux system
CUDA: 8.0 and CuDNN 6.0.
NVIDIA Driver: Latests on a GTX 1080.

**Comment by raver119** [comment]
Any chance for serialized model that reproduces this behavior?

**Comment by sorensf** [comment]
Yes, but I'm still working on it sorry. So far the stand-alone examples that I've tried making to reproduce the behavior doesn't show the issue, even when using the existing serialized model I have.

**Comment by raver119** [comment]
That's even more interesting then.

**Comment by sorensf** [comment]
I'm running a setup where I have 3 NN models loaded in memory within a larger system. When switching between CPU and GPU the third model on GPU changes its output from "expected" to something close to an untrained network. At the moment when I just try loading the 3 models within a small test, no change in behavior is exhibited between CPU and GPU...

**Comment by raver119** [comment]
Are they used in the same thread? If yes - that might be caused by workspaces. 
It's rather simple to check though: set inference mode to NONE.

**Comment by sorensf** [comment]
They are, but I have tried with inference mode NONE and unfortunately didn't help.

**Comment by raver119** [comment]
What happens if you load only 1 model?

**Comment by sorensf** [comment]
I'm at home at the moment, and I haven't tested that yet, but will do it first thing tomorrow.

**Comment by raver119** [comment]
Any news?

**Comment by sorensf** [comment]
I managed to reproduce the issue in a unit test (note that it uses some helper methods): https://gist.github.com/sorensf/667a944b41a4a631d56491af17c7b561

Talking with Raver119 it appears that the problem is that when running on CUDA backend, when one thread deserializes a network it does not actually complete the deserialization before another operation is performed on the network in that thread, so if another thread uses that network it just appears as an untrained network. 

Calling Nd4j.getExecutioner().commit() or .output(..) in the deserializing thread serves as a workaround for the issue.

**Comment by sorensf** [comment]
Source chat history: https://gist.github.com/sorensf/10d8da81568914cd6c80372d9c4872a2

**Comment by sorensf** [comment]
Another observation, it seems that if I use the model returned from ModelSerializer.restoreMultiLayerNetwork(...) directly, then the issue also disappears. But when I use the multiLayerNetwork.getLayerWiseConfigurations() multiLayerNetwork.params() to create a new network, that network has the issue.
So the problem does not seem to be the deserialization but rather the copying of parameters. (My deserialization code https://gist.github.com/sorensf/23ee692101449360bf136158cbcda8cb)
